[API] Support `zero_infinity` in `ctc_loss`

aztice opened the pull request:
### PR Category
User Experience

### PR Types
New features

### Description
为 `ctc_loss` 添加 `zero_infinity` 参数

aztice committed:
feat: ctc_loss.zero_infinity

aztice committed:
fix: code-style issue.

aztice committed:
fix: code-style issue.

zxcd commented on python/paddle/nn/functional/loss.py:
```diff
@@ -2042,6 +2044,11 @@ def warpctc(
+            paddle.isinf(loss_out), paddle.zeros_like(loss_out), loss_out
+        )
+
     loss_out = paddle.squeeze(loss_out, [-1])
```
whether add zero_infinity under squeeze to avoid useless compute?

aztice replied:
thanks for your advise, modified~

aztice committed:
optimize: reduce calculation

aztice committed:
fix: code-style issue.

aztice committed:
feat: test_warpctc_zero_infinity

aztice committed:
fix: code-style issue.

aztice committed:
fix: code-style issue.

aztice committed:
fix: zero_infinity for ctc_loss

aztice committed:
fix: code-style issue.

aztice committed:
fix: code-style issue.

aztice commented:
/re-run all-failed

zxcd reviewed (approved):
LGTM

jeff41404 commented:
shall we also support `zero_infinity` in API [paddle.nn.CTCLoss](https://www.paddlepaddle.org.cn/documentation/docs/zh/develop/api/paddle/nn/CTCLoss_cn.html#ctcloss) to keep consistency of loss function?

aztice committed:
feat: ctcloss.zero_infinity

aztice commented:
@jeff41404 PTAL thx nya~

aztice commented:
/re-run all-failed

aztice committed:
fix: docs issue.

aztice committed:
fix: error

aztice committed:
fix: code-style issue.

aztice committed:
fix: errors

aztice committed:
Fix loss output handling for zero_infinity case

aztice committed:
fix: errors

aztice committed:
Replace zeros_like with full_like for loss output

aztice committed:
Refactor handling of infinite loss values

aztice committed:
Improve readability of loss_out assignment

Refactor loss_out assignment for better readability.

aztice committed:
Fix paddle.where parameters in loss function

```
W1028 22:18:12.514281 25474 eager_utils.cc:3441] Paddle static graph(PIR) not support input out tensor for now!!!!!
I1028 22:18:18.986547 25474 pir_interpreter.cc:1524] New Executor is Running ...
W1028 22:18:18.987156 25474 gpu_resources.cc:114] Please NOTE: device: 0, GPU Compute Capability: 7.0, Driver API Version: 12.0, Runtime API Version: 11.8
I1028 22:18:18.988493 25474 pir_interpreter.cc:1547] pir interpreter is running by multi-thread mode ...
.
----------------------------------------------------------------------
Ran 1 test in 6.494s

OK
```

jeff41404 reviewed (dismissed):
LGTM

luotao1 reviewed (commented):
@aztice 请对应修改中文文档

aztice commented:
> @aztice 请对应修改中文文档

明白。

aztice commented:
### Docs UPDATE
> https://github.com/PaddlePaddle/docs/pull/7597

SigureMo commented on python/paddle/nn/layer/loss.py:
```diff
@@ -1352,6 +1352,7 @@ def forward(
         input_lengths: Tensor,
         label_lengths: Tensor,
         norm_by_times: bool = False,
+        zero_infinity: bool = False,
```
这边文档也需要补充

aztice replied:
完成。

SigureMo commented on test/legacy_test/test_layers.py:
```diff
@@ -1107,6 +1107,57 @@ def test_warpctc_with_padding(self):
             )
             return output
 
+    def test_ctc_loss_zero_infinity(self):
```
不应该写在这，应该写在 test/legacy_test/test_warpctc_op.py

而且单测应该优先测动态图，当然最好同时测静态图

aztice replied:
明白。

aztice committed:
Remove test_ctc_loss_zero_infinity function

Removed the test for CTC loss with zero and infinity values.

aztice committed:
Add tests for CTC loss with zero_infinity

Added tests for CTC loss with zero_infinity flag in both static and dynamic graph modes.

aztice committed:
Format labels_np initialization for readability

aztice committed:
Refactor string literals and improve formatting

aztice committed:
Improve code formatting in test_warpctc_op.py

Refactor code for better readability by adjusting line breaks and formatting in test_warpctc_op.py.

aztice committed:
Refactor static_graph context manager usage

aztice committed:
Use fixed seed value in static_graph context manager

Replaced the use of self.seed with a fixed seed value for consistency in tests.

SigureMo commented on test/legacy_test/test_warpctc_op.py:
```diff
@@ -754,6 +755,99 @@ def test_functional_api():
+                paddle.framework.random._manual_program_seed(seed)
+            else:
+                paddle.framework.random._manual_program_seed(seed)
+            yield
```
这个函数和静态图有什么关系么？为什么叫 `static_graph`？而且什么时候会走到非 PIR 分支？

aztice replied:
as i recalled, i got it from `test/legacy_test/test_layers.py` instead which means i copied it again. i will modify it.

aztice committed:
Refactor static graph usage in CTC loss test

Refactored test to use paddle.enable_static() and removed context manager for static graph.

SigureMo commented on test/legacy_test/test_warpctc_op.py:
```diff
@@ -766,57 +766,46 @@ def test_ctc_loss_zero_infinity(self):
         input_len_np = np.array([1], dtype=np.int64)
         label_len_np = np.array([3], dtype=np.int64)
```
```python
main_program = paddle.static.Program()
startup_program = paddle.static.Program()
with paddle.static.program_guard(main_program, startup_program):
    # 组网逻辑

exe.run(startup_program) # 如果需要，运行 startup program，但大概率你的 case 并不需要
out, = exe.run(
    main_program,
    feed=...,
    fetch_list=...,
)

# 用 np.testing.assert_allclose(...) 检查
```

用 `program_guard` 避免对全局 main program 的修改，类似参考代码有很多

SigureMo replied:
为什么每次 codestyle 都不过？没参考贡献指南配置 `pre-commit` 么？https://www.paddlepaddle.org.cn/documentation/docs/zh/dev_guides/code_contributing_path_cn.html

aztice replied:
上次安precommit我记得自动安装paddle了，卡住了然后如果每次都要precommit需要大量时间，目前用的ruff来解决这个问题，大多代码都挺适用，不过平时有时候不在家我就得用手机改就容易出现codestyle error

aztice replied:
完成

aztice committed:
Refactor CTC loss test to use program guard

aztice committed:
Remove unused import in test_warpctc_op.py

Removed unused import of contextlib.

aztice committed:
Enable and disable static mode in warpctc test

aztice commented:
/re-run all-failed

SigureMo commented on test/legacy_test/test_warpctc_op.py:
```diff
@@ -754,6 +754,93 @@ def test_functional_api():
+            )[0]
+
+            # illegal sample -> 0
+            self.assertTrue(np.allclose(loss_val[0], 0.0, atol=1e-6))
```
> `# 用 np.testing.assert_allclose(...) 检查`

```
****************
0. It is recommended to use 'np.testing.assert_allclose' and 'np.testing.assert_array_equal' instead of 'self.assertTrue(np.allclose(...))' and 'self.assertTrue(np.array_equal(...))'.
Please modify the code below. If anything is unclear, please read the specification [ https://github.com/PaddlePaddle/community/blob/master/rfcs/CodeStyle/20220805_code_style_improvement_for_unittest.md#background ]. If it is a mismatch, please request SigureMo (Recommend) or zrr1999 review and approve.
The code that do not meet the specification are as follows:
+            self.assertTrue(np.allclose(loss_val[0], 0.0, atol=1e-6))

There are 1 approved errors.
****************
```

不是特意说了用 `np.testing.assert_allclose` 了么？

aztice replied:
明白

SigureMo commented on test/legacy_test/test_warpctc_op.py:
```diff
@@ -754,6 +754,93 @@ def test_functional_api():
+            blank=n_class - 1,
+        )
+
+        np.testing.assert_allclose(loss.numpy()[0], 0.0, rtol=1e-6)
```
```suggestion
        np.testing.assert_allclose(loss.numpy(), [0.0], rtol=1e-6)
```

shape 也是检查项

aztice committed:
Update test/legacy_test/test_warpctc_op.py

Co-authored-by: Nyakku Shigure <sigure.qaq@gmail.com>

aztice committed:
Replace assertTrue with assert_allclose in tests

SigureMo commented on test/legacy_test/test_warpctc_op.py:
```diff
@@ -754,6 +754,94 @@ def test_functional_api():
+            )[0]
+
+            # illegal sample -> 0
+            np.testing.assert_allclose(loss_val[0], 0.0, atol=1e-6)
```
```suggestion
            np.testing.assert_allclose(loss_val, [0.0], atol=1e-6)
```

SigureMo committed:
Update test/legacy_test/test_warpctc_op.py

aztice commented:
/re-run all-failed

SigureMo commented:
我刚刚没注意，`CTCLoss` 的实现有问题吧？为什么新增参数 `zero_infinity` 是作为 `forward` 输入？torch 是 `__init__` 的输入吧？？？

aztice commented:
> 我刚刚没注意，`CTCLoss` 的实现有问题吧？为什么新增参数 `zero_infinity` 是作为 `forward` 输入？torch 是 `__init__` 的输入吧？？？

稍等我看看

aztice commented:
> 我刚刚没注意，`CTCLoss` 的实现有问题吧？为什么新增参数 `zero_infinity` 是作为 `forward` 输入？torch 是 `__init__` 的输入吧？？？

确认了下，稍等我开个新PR改下，`forward` 以及 `__init__` 均支持输入。
